MultiCoupons/Controller/Cart/CouponPost.php

```diff
@@ -0,0 +1,183 @@
+
+        if (!$itemsCount) {
+            return $this->_goBack();
+        }
+
+        $removeCoupons = $this->getRequest()->getParam('remove') ?: [];
+        $couponCodes = $this->getRequest()->getParam('coupon_code') ?: [];
+
+        $resultCoupons = $this->getCouponCodes($cartQuote, $couponCodes, $removeCoupons);
+
+        if ($resultCoupons) {
+            try {
+                $cartQuote->getShippingAddress()->setCollectShippingRates(true);
+                $cartQuote->setCouponCode($resultCoupons)->collectTotals();
+                $this->_checkoutSession->getQuote()->setCouponCode($resultCoupons)->save();
+                $this->messageManager->addSuccess(
+                    __(
+                        'You used coupon code "%1".',
+                        $this->_objectManager
+                            ->get(Escaper::class)
+                            ->escapeHtml($resultCoupons)
+                    )
+                );
+            } catch (\Exception $e) {
+                $this->messageManager->addError($e->getMessage());
+            }
+        } else {
+            $this->messageManager->addSuccess(__('You canceled the coupon code.'));
+        }
+
+        return $this->_goBack();
+    }
+
+    /**
+     * @param MagentoQuote $cartQuote
+     * @param array $couponCodes
+     * @param array $removeCoupons
+     * @return string
+     */
+    public function getCouponCodes(
+        MagentoQuote $cartQuote,
+        array $couponCodes,
+        array $removeCoupons
+    ): string {
+        $oldCouponCode = $cartQuote->getCouponCode();
+
+        if ($oldCouponCode) {
+            $arrayOldCouponCodes = explode(',', $oldCouponCode);
+        } else {
+            $arrayOldCouponCodes = [];
+        }
+
+        $validatedCodes = [];
+        foreach ($couponCodes as $code) {
+            if ($code && $this->isValidCouponCode($code)) {
+                $validatedCodes[] = $code;
+            }
+        }
+
+        $arrayOldCouponCodes = array_diff($arrayOldCouponCodes, $removeCoupons);
+        $arrayNewCoupons = array_diff($validatedCodes, $arrayOldCouponCodes);
+        $resultCoupons = implode(',', array_merge($arrayOldCouponCodes, $arrayNewCoupons));
+
+        if ($oldCouponCode == $resultCoupons) {
+            return '';
+        }
+
+        return $resultCoupons;
+    }
+
+    /**
+     * @param string $code
+     * @return bool
+     */
+    public function isValidCouponCode(string &$code): bool
+    {
+        if (!$code) {
+            return false;
+        }
+
+        $code = trim($code);
+        $codeLength = strlen($code);
+
+        $coupon = $this->couponFactory->create();
+
+        $coupon->load($code, 'code');
+        if (
+            $codeLength &&
+            $codeLength <= CartHelper::COUPON_CODE_MAX_LENGTH &&
+            $coupon->getId()
+        ) {
+            return true;
+        } else {
+            $this->messageManager->addError(
```

Comment:
В идеале - избегать else и излишней вложенности, например:
```suggestion
        if (
            $codeLength &&
            $codeLength <= CartHelper::COUPON_CODE_MAX_LENGTH &&
            $coupon->getId()
        ) {
            return true;
        }
        
        ...->addError(...);
        
        return false;
```